In [1]:
from scipy.optimize import linprog
import numpy as np

Objective function:
$$z(max) = 7x_1 + 6x_2$$

Constraints:
$$C_1 = 2*x_1 + 3*x_2 \leq 12$$
$$C_2 = 6*x_1 + 5*x_2 \leq 30$$
$$x_1,x_2 \geq 0$$

In [2]:
z = np.array([7,6])

C for constraints, which is going to store all constraints of the problem.

In [3]:
C = np.array([
    [2,3],            #C1
    [6,6]             #C2
])

Then store the 12 and 30 for each constraint.

In [4]:
b = np.array([12,30])

To solve the problem we need to call linprog, and store the solution in a variable. 

We are going to use sol.

In [5]:
sol = linprog(-z, A_ub = C, b_ub = b, bounds = (0,None))

In [6]:
print(sol)

     fun: -35.0
 message: 'Optimization terminated successfully.'
     nit: 1
   slack: array([ 2.,  0.])
  status: 0
 success: True
       x: array([ 5.,  0.])


To print the values that are needed we are going to use:

In [7]:
print("x1 = {}, x2 = {}, z = {}".format(sol.x[0], sol.x[1], sol.fun*-1))

x1 = 5.0, x2 = 0.0, z = 35.0


Using an easy way by multiplying our z variable (z = np.array([7,6])) and our sol.x values (sol.x).

In [8]:
OF = z*sol.x
print("z = {}".format(OF.sum()))

z = 35.0


Our LP problem is solved now we are going to replace those values into the objective function.
$$z(max) = 7x_1 + 6x_2$$
$$z(max) = 7*5 + 6*0$$
$$z(max) = 35$$

Let's solve a LP more complex: 

*traducir la wea :v*

Defina $X_1$ como el numero de mesas, $X_2$ como el numero de sofas y $X_3$ como el numero de sillas que se producirian cada mes. La utilidad se calcula como el ingreso por cada articulo menos el costo de materiales(madera y tela), menos el costo de mano de obra. Dado que la mano de obra es fija, se resta como una cantidad total. En terminos matematicos se tiene $(400 - 100)X_1 + (750 - 75 - 175)X_2 + (240 - 40)X_3 - 75000$ . La utilidad se calcula asi:

$$Utilidad = 300X_1 + 500X_2 + 200X_3 - 75000$$

Las restricciones son las siguientes:

1. Madera: $10X_1 + 7.5X_2 + 4X_3 \leq 4350$
2. Tela: $10X_2 \leq 2500$
3. Serrado: $0.5X_1 + 0.4X_2 + 0.5X_3 \leq 280$
4. Cortado: $0.4X_2 \leq 140$
5. Lijado:  $0.5X_1 + 0.1X_2 + 0.5X_3 \leq 280$
6. Entintado: $0.4X_1 + 0.2X_2 + 0.4X_3 \leq 140$
7. Montaje: $1X_1 + 1.5X_2 + 0.5X_3 \leq 700$

Demanda:

8. Mesas: $X_1 \leq 300$
9. Sofas: $X_2 \leq 180$
10. Sillas: $X_3 \leq 400$

In [9]:
z = np.array([300, 500, 200])
C = np.array([
    [ 10, 7.5,   4], #C1
    [  0,  10,   0], #C2
    [0.5, 0.4, 0.5], #C3
    [  0, 0.4,   0], #C4
    [0.5, 0.1, 0.5], #C5
    [0.4, 0.2, 0.4], #C6
    [  1, 1.5, 0.5], #C7
    [  1,   0,   0], #C8
    [  0,   1,   0], #C9
    [  0,   0,   1]  #C10
])

b = np.array([4350, 2500, 280, 140, 280, 140, 700, 300, 180, 400])

In [10]:
sol = linprog(-z, A_ub = C, b_ub = b, bounds=(0,None))

In [11]:
print(sol)

     fun: -168000.0
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([ 400.,  700.,   78.,   68.,  132.,    0.,  170.,   40.,    0.,  400.])
  status: 0
 success: True
       x: array([ 260.,  180.,    0.])


In [12]:
print('X1 = {}, X2 = {}, X3 = {}'.format(sol.x[0], sol.x[1], sol.x[2]))

X1 = 260.0, X2 = 180.0, X3 = 0.0


For $X_1, X_2, X_3$ in:
$$ Utilidad = 300X_1 + 500X_2 + 200X_3 - 75 000 $$
$$ Utilidad = 300*260 + 500*180 + 200*0 - 75 000$$
$$ Utilidad = 93 000$$
Solving in python:

In [13]:
print('Utilidad = 300*X1 + 500*X2 + 200X3 - 75000\n')

OF = z*sol.x
print('Utilidad = {}'.format(OF.sum()-75000))

Utilidad = 300*X1 + 500*X2 + 200X3 - 75000

Utilidad = 93000.0
